### PR TITLE
[SPARK-6893][ML] default pipeline parameter handling in python

### PR DESCRIPTION
**Pull request by mengxr**
Same as #5431 but for Python. @jkbradley

### CODE REVIEW

File: python/pyspark/ml/param/__init__.py [hunk 1]
```diff
@@ -52,10 +50,11 @@ class Params(Identifiable):
 
     __metaclass__ = ABCMeta
 
-    def __init__(self):
-        super(Params, self).__init__()
-        #: embedded param map
-        self.paramMap = {}
+    #: internal param map for user-supplied values param map
+    paramMap = {}
```
Comment (jkbradley) [inline python/pyspark/ml/param/__init__.py]: Curious: Why remove the init method and the call to ```Identifiable.__init__```?

File: python/pyspark/ml/param/__init__.py [hunk 2]
```diff
@@ -25,23 +25,21 @@
...
+        if not isinstance(parent, Params):
+            raise ValueError("Parent must be a Params but got type %s." % type(parent).__name__)
         self.parent = parent
         self.name = str(name)
         self.doc = str(doc)
-        self.defaultValue = defaultValue
 
     def __str__(self):
-        return str(self.parent) + "-" + self.name
+        return str(self.parent) + "__" + self.name
```
Comment (jkbradley) [inline python/pyspark/ml/param/__init__.py]: Is this important?  Should it be tested somewhere?

File: python/pyspark/ml/param/__init__.py [hunk 3]
```diff
@@ -52,10 +50,11 @@ class Params(Identifiable):
...
-        #: embedded param map
-        self.paramMap = {}
+    #: internal param map for user-supplied values param map
+    paramMap = {}
+
+    #: internal param map for default values
+    defaultParamMap = {}
 
     @property
     def params(self):
```
Comment (jkbradley) [inline python/pyspark/ml/param/__init__.py]: I should have done this previously, but can these be sorted by name (as in Scala)?

File: python/pyspark/ml/param/__init__.py [hunk 4]
```diff
@@ -67,11 +66,112 @@ def params(self):
...
+        elif isinstance(param, str):
+            return self.getOrDefault(self.getParam(param))
+        else:
+            raise KeyError("Cannot recognize %r as a param." % param)
+
+    def extractParamMap(self, extraParamMap={}):
+        """
+        Extracts the embedded default param values and user-supplied
+        values, and then merges them with extra values from input into
+        a flat param map, where the latter values is used if there
```
Comment (jkbradley) [inline python/pyspark/ml/param/__init__.py]: "is" --> "are"

File: python/pyspark/ml/param/__init__.py [hunk 5]
```diff
@@ -67,11 +66,112 @@ def params(self):
...
+        Validates that the input param belongs to this Params instance.
+        """
+        if param.parent is not self:
+            raise ValueError("Param %r does not belong to %r." % (param, self))
+
+    def _resolveParam(self, param):
+        """
+        Resolves a param and validates the ownership.
+        :param param: param name or the param instance, which must
+                      belongs to this Params instance
```
Comment (jkbradley) [inline python/pyspark/ml/param/__init__.py]: "belongs" --> "belong"

File: python/pyspark/ml/util.py [hunk 6]
```diff
@@ -41,7 +41,7 @@ class Identifiable(object):
     def __init__(self):
         #: A unique id for the object. The default implementation
         #: concatenates the class name, "-", and 8 random hex chars.
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: Update doc: "-" --> "_"

**Comment by jkbradley** [comment]
@mengxr  Looks good other than those minor comments above

File: python/pyspark/ml/param/__init__.py (hunk 1 above)
Comment (mengxr) [inline python/pyspark/ml/param/__init__.py]: That's the default behavior. `paramMap` doesn't need reference to `self`, so this is simpler.

File: python/pyspark/ml/param/__init__.py (hunk 2 above)
Comment (mengxr) [inline python/pyspark/ml/param/__init__.py]: This is to match sklearn's approach to specify parameters. This is just for display at this time.

File: python/pyspark/ml/param/__init__.py (hunk 3 above)
Comment (mengxr) [inline python/pyspark/ml/param/__init__.py]: I use `dir` here, which returns a sorted list.

**Comment by mengxr** [comment]
Merged into master.

**Pull request closed without merging**